pytorch_lightning/plugins/collective/collective_plugin.py

```diff
@@ -0,0 +1,56 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from abc import ABC, abstractmethod
+from typing import Any, List, Optional, Union
+
+import torch
+
+from pytorch_lightning.utilities.distributed import ReduceOp
+
+
+class Collective(ABC):
+    """Base class for collective functions for training type plugins."""
+
+    @abstractmethod
+    def barrier(self, name: Optional[str] = None) -> None:
+        """Forces all possibly joined processes to wait for each other."""
+
+    @abstractmethod
+    def broadcast(self, obj: object, src: int = 0) -> object:
```

Comment:
Looking at the derived classes, this really should be called `broadcast_object`, which has a different semantic from the plain `broadcast` for tensors.

For example, when NCCL `broadcast` returns (or `.wait()` on the async handle returns), it is only guaranteed that the operation is enqueued to a non-default Cuda stream. The main thread will only block when the result (or something that is dependent on the result) is used to produce side effect (e.g. printing the result).

However, since `broadcast_object_list` (or its variant for other accelerator types) involves serialization/deserialization, the main thread will always block before the collective finishes. This can lead to performance implications when `broadcast_object_list` is mis-used to transfer large tensors.

Reply:
Should we go through a depreciation process and provide a broadcast_object hook ?

Reply:
@tchaton yes I think distinguishing `broadcast` and `broadcast_object(_list])` in the `Collective` interface would be nice.

Reply:
collective intent to hide behavior of different backend and have a unified api. Different collective also have different broadcast, broadcast_object, broadcast_object_list names, I feel it's hard to choose a specific one.
For example:
1. torch collective : version < 1.8 using broadcast
                              version >= 1.8 using broadcast_object_list
2. horovod: broadcast_object  (official document mentioned it's mainly for broadcast optimizer.state_dict(), for model param has broadcast_parameters? https://horovod.readthedocs.io/en/stable/api.html#module-horovod.torch)

 Does  broadcast vs broadcast_object vs broadcast_object_list name really matter here? How about we add more detailed comments?

Reply:
> Does broadcast vs broadcast_object vs broadcast_object_list name really matter here? How about we add more detailed comments?

as @yifuwang points out, there are performance considerations with broadcasting large tensors depending on the underlying library call used. we could either:
- offer dedicated `broadcast` vs `broadcast_object_list` interfaces on the collective
- special-case the implementation of `broadcast` on collective implementations based on the type of the input passed in. 

of these i prefer the first option since it's the most explicit

Reply:
>offer dedicated broadcast vs broadcast_object_list interfaces on the collective

Current broadcast usage is 
1. in timer.py -> broadcast(should stop)
2. in model_checkpoint.py -> broadcast(skip_time), broadcast(ckpt_path), broadcast(exists)
3. in Trainer properties -> broadcast(dirpath)

I don't think having both broadcast and broadcast_object_list two API fit our use cases because:
1. Which one should we use in the three callers mentioned above?
2. Not all backend has broadcast and broadcast_object/broadcast_object_list. Like hovorod only have broadcast_object vs TPUspawn only have broadcast lightning defined without using any backend

Reply:
sorry I was wrong. Horovod has broadcast (totally 7 different types of broadcast)